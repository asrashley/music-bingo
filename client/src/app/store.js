import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import { connectRouter } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import { reducer as formReducer } from 'redux-form';

import cardsReducer from '../cards/cardsSlice';
import gamesReducer from '../games/gamesSlice';
import ticketsReducer from '../tickets/ticketsSlice';
import userReducer from '../user/userSlice';
import adminReducer from '../admin/adminSlice';
import usersMiddleware from '../user/userMiddleware';

export const history = createBrowserHistory();

export const store = configureStore({
  middleware: [...getDefaultMiddleware(), usersMiddleware, logger],
  reducer: {
    admin: adminReducer,
    cards: cardsReducer,
    form: formReducer,
    games: gamesReducer,
    tickets: ticketsReducer,
    router: connectRouter(history),
    user: userReducer,
  },
});
