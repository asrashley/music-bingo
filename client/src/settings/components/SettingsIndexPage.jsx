import React, { useEffect } from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { reverse } from 'named-urls';

/* actions */
import { fetchUserIfNeeded } from '../../user/userSlice';
import { fetchSettingsIfNeeded } from '../../settings/settingsSlice';

/* selectors */
import { getUser } from '../../user/userSelectors';
import { getSettingsSections } from '../../settings/settingsSelectors';

/* data */
import { routes } from '../../routes/routes';

import '../styles/settings.scss';

export function SettingsIndexPage() {
  const dispatch = useDispatch();
  const user = useSelector(getUser);
  const settingsSections = useSelector(getSettingsSections);

  useEffect(() => {
    dispatch(fetchUserIfNeeded());
  }, [dispatch]);

  useEffect(() => {
    if (user.loggedIn) {
      dispatch(fetchSettingsIfNeeded('all'));
    }
  }, [dispatch, user]);

  if (user.groups.admin !== true) {
    return (
      <div id="settings-page">
        <div className="alert alert-warning" role="alert">
          <span className="error-message">Only an admin can modify settings</span>
        </div>
      </div>
    );
  }

  return (
    <div id="settings-page">
      <div className="user-commands">
        {settingsSections.map((section) => <Link
          className="btn btn-lg btn-primary mb-4"
          to={reverse(`${routes.settingsSection}`, { section })}
          key={section}
        >Modify {section} Settings
        </Link>)}
      </div>
    </div >
  );
}
