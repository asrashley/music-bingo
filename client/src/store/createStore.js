import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import { connectRouter } from 'connected-react-router';

import adminReducer from '../admin/adminSlice';
import directoriesReducer from '../directories/directoriesSlice';
import gamesReducer from '../games/gamesSlice';
import messagesReducer from '../messages/messagesSlice';
import settingsReducer from '../settings/settingsSlice';
import systemReducer from '../system/systemSlice';
import ticketsReducer from '../tickets/ticketsSlice';
import userReducer from '../user/userSlice';
import usersMiddleware from '../user/userMiddleware';
import messagesMiddleware from '../messages/messagesMiddleware';

import { history } from './history';

export function createStore(preloadedState, logging) {
  const middleware = [...getDefaultMiddleware(), usersMiddleware, messagesMiddleware];
  if (logging === undefined) {
    logging = (process.env.NODE_ENV === 'development');
  }
  if (logging) {
    middleware.push(logger);
  }
  return configureStore({
    middleware,
    reducer: {
      admin: adminReducer,
      directories: directoriesReducer,
      games: gamesReducer,
      messages: messagesReducer,
      settings: settingsReducer,
      system: systemReducer,
      tickets: ticketsReducer,
      router: connectRouter(history),
      user: userReducer,
    },
    preloadedState
  });
}
