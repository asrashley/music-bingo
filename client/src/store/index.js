export { createStore } from './createStore';
export { initialState } from './initialState';
export { history } from './history';
