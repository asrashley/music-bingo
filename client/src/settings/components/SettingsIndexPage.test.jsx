import React from 'react';
import fetchMock from 'fetch-mock';

import { renderWithProviders } from '../../../tests';
import { MockBingoServer, adminUser, normalUser } from '../../../tests/MockServer';
import { SettingsIndexPage } from './SettingsIndexPage';
import { initialState, createStore } from '../../store';

describe('SettingsIndexPage component', () => {
  const settingsSections = ['app', 'database', 'privacy', 'smtp'];
  let apiMock;

  beforeEach(() => {
    apiMock = new MockBingoServer(fetchMock);
  });

  afterEach(() => {
    apiMock = null;
    fetchMock.reset();
  });

  afterAll(() => {
    vi.restoreAllMocks();
  });

  it('only an Admin can modify settings', async () => {
    const store = createStore({
      ...initialState,
      user: apiMock.getUserState(normalUser),
    });
    const { getByText, queryByText } = renderWithProviders(<SettingsIndexPage />, { store });
    getByText('Only an admin can modify settings');
    settingsSections.forEach(section =>
      expect(queryByText(section, { exact: false })).toBeNull());
  });

  it('shows link to each settings section', async () => {
    const store = createStore({
      ...initialState,
      user: apiMock.getUserState(adminUser),
    });
    const { findByText, asFragment } = renderWithProviders(<SettingsIndexPage />, { store });
    await Promise.all(settingsSections.map(async (section) => {
      const link = await findByText(`Modify ${section} Settings`);
      expect(link.href).toMatch(new RegExp(`/user/settings/${section}$`));
    }));
    expect(asFragment()).toMatchSnapshot();
  });
});
