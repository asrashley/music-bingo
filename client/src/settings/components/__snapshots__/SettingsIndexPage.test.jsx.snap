// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SettingsIndexPage component > shows link to each settings section 1`] = `
<DocumentFragment>
  <div
    id="settings-page"
  >
    <div
      class="user-commands"
    >
      <a
        class="btn btn-lg btn-primary mb-4"
        href="/user/settings/app"
      >
        Modify app Settings
      </a>
      <a
        class="btn btn-lg btn-primary mb-4"
        href="/user/settings/database"
      >
        Modify database Settings
      </a>
      <a
        class="btn btn-lg btn-primary mb-4"
        href="/user/settings/privacy"
      >
        Modify privacy Settings
      </a>
      <a
        class="btn btn-lg btn-primary mb-4"
        href="/user/settings/smtp"
      >
        Modify smtp Settings
      </a>
    </div>
  </div>
</DocumentFragment>
`;
